src/move/emojicoin_dot_fun/sources/tests.move

```diff
@@ -1469,6 +1508,400 @@
+                base_volume = b_out - pool_fee_base;
+                cpamm_real_reserves = MockReserves {
+                    base: cpamm_real_reserves.base - base_volume,
+                    quote: cpamm_real_reserves.quote + quote_volume,
+                };
+            } else { // Buy against bonding curve, might require a state transition.
+                let quote_volume_before_state_transition =
+                    get_QUOTE_VIRTUAL_CEILING() - clamm_virtual_reserves.quote;
+                // If a state transition, need to swap against CLAMM then CPAMM.
+                if (quote_volume >= quote_volume_before_state_transition) {
+                    results_in_state_transition = true;
+                    base_volume = clamm_virtual_reserves.base - get_BASE_VIRTUAL_FLOOR();
+                    let remaining_quote_volume =
+                        quote_volume - quote_volume_before_state_transition;
+                    cpamm_real_reserves = if (remaining_quote_volume > 0) {
+                        let b_out_cpamm = cpamm_simple_swap_output_amount(
+                            remaining_quote_volume,
+                            SWAP_BUY,
+                            pack_mock_reserves(base_cpamm_real_reserves_exact_transition())
+                        );
+                        pool_fee_base = get_bps_fee(b_out_cpamm, get_POOL_FEE_RATE_BPS());
+                        let base_volume_cpamm = b_out_cpamm - pool_fee_base;
+                        base_volume = base_volume + base_volume_cpamm;
+                        MockReserves {
+                            base: base_cpamm_real_reserves_exact_transition().base -
+                                base_volume_cpamm,
+                            quote: base_cpamm_real_reserves_exact_transition().quote +
+                                remaining_quote_volume
+                        }
+                    } else { // Exact state transition
+                        base_cpamm_real_reserves_exact_transition()
+                    };
+                    clamm_virtual_reserves = MockReserves { base: 0, quote: 0 };
+                } else { // Buy against bonding curve, no state transition.
+                    base_volume = cpamm_simple_swap_output_amount(
+                        quote_volume,
+                        SWAP_BUY,
+                        pack_mock_reserves(setup_market_view.clamm_virtual_reserves)
+                    );
+                    clamm_virtual_reserves = MockReserves {
+                        base: clamm_virtual_reserves.base - base_volume,
+                        quote: clamm_virtual_reserves.quote + quote_volume,
+                    };
+                }
+            };
+            net_proceeds = base_volume;
+            market_aptos_coin_balance = setup_market_view.aptos_coin_balance + quote_volume;
+            market_emojicoin_balance = setup_market_view.emojicoin_balance - base_volume;
+        };
```

Comment:
Is this all duplicated logic from `simulate_swap_inner` just to recreate the mock data? I think all of the duplicated logic here just for the test setup is a sign that the function logic should be broken up

I mean I understand that the unit test framework isn't great for test setups and stuff, but it's difficult to tell here what exactly is being tested vs what's being duplicated

Reply:
> Is this all duplicated logic from `simulate_swap_inner`

Not copy-pasted, but yes the same logic (I reimplemented without looking at
source code, then verified that both had same output)

The catch is that without something like this, I'd be writing 9 or so individual
tests with a bunch of the same code blocks. Then if something in the core logic
changed, we'd have to update the manual code for each test, so not very
maintainable.

I'm trying to get enough abstraction to ease testing, but still to have the
testing be simple enough to ensure that the verification is happening in a
straightforward manner. Hence I agree with:

> sign that the function logic should be broken up

In the ideal case, I think we would:

1. Write core source code in a maximally-abstracted manner, then test all the
   individual helper functions on their own.
1. Write tests to use a central harness flow like the one I've added here, but
   rely mainly on the already-verified helper functions.
1. Test everything to 100% coverage with full state mutation assertions on each
   test.
1. Manually optimize/inline the core code, while ensuring tests still pass.

Unfortunately, it looks like we went straight to the final step rather than
optimizing at the end, so there aren't as many of these decomposed helpers
available for testing.

I'll start implementing some and see how we can make the test harness more
elegant.

Reply:
Per eng call updated to

https://github.com/econia-labs/emojicoin-dot-fun/blob/75909f311ce684031191738f648fe06c4af01b1a/src/move/emojicoin_dot_fun/sources/tests.move#L1611-L1640

Reply:
Thanks for doing this for lil old me, it's significantly easier for me to read and understand 😄